cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+        {
+            if (package_data->arch)
+            {
+                /* Some error occurred as we already have arch for 
+                 * given package. */
+               Log(LOG_LEVEL_ERR, "Duplicated package architecture "
+                    "received: [%s] %s", line, package_data->arch);
+                free(package_data);
+                return NULL;
+            }
+            package_data->arch = 
+                SafeStringDuplicate(line + strlen("Architecture="));
+        }
+        /* For handling errors */
+        else if (StringStartsWith(line, "Error="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "have error: %s",
+                line + strlen("Error="));
+        }
+        else if (StringStartsWith(line, "ErrorMessage="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "have error message: %s",
+                line + strlen("ErrorMessage="));
+        }
+        else
+        {
+            Log(LOG_LEVEL_VERBOSE, "unsupported option: %s", line);
+        }
+    }
+    
+    return package_data;
+}
+
+static int NegotiateSupportedAPIVersion(PackageModuleWrapper *wrapper)
+{
+    int api_version = -1;
+
+    Rlist *response = NULL;
+    if (ReadWriteDataToPackageScript("supports-api-version", "",
+                                     &response, wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "wrapper while getting supported API version.");
+        return -1;
+    }
+    
+    if (response)
+    {
+        if (RlistLen(response) == 1)
+        {
+            api_version = atoi(RlistScalarValue(response));
+            Log(LOG_LEVEL_DEBUG, "package wrapper API version: %d", api_version);
+        }
+        RlistDestroy(response);
+    }
+    return api_version;
+}
+
+/* IMPORTANT: this might not return all the data we need like version
+              or architecture but package name MUST be known. */
+static
+PackageInfo *GetPackageData(const char *name, Rlist *options,
+                            const PackageModuleWrapper *wrapper)
+{   
+    char *options_str = ParseOptions(options);
+    char *request = StringFormat("%sFile=%s\n", options_str, name);
+    
+    Rlist *response = NULL;
+    if (ReadWriteDataToPackageScript("get-package-data", request, &response,
+            wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "package module while collecting package data.");
+        free(options_str);
+        free(request);
+        return NULL;
+    }
+    
+    PackageInfo *package_data = NULL;
+    
+    if (response)
+    {   
+        package_data = ParseAndCheckPackageDataReply(response);
+        RlistDestroy(response);
+        
+        if (package_data)
+        {
+            /* At this point at least package name MUST be known (if no error) */
+            if (!package_data->name || package_data->type == PACKAGE_TYPE_NONE)
```

Comment:
Difference between comment and code: the comment demands name, the code demands name **and type**.
The comment should probably `s/name/name and type/`